core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala

```diff
@@ -614,6 +585,74 @@ private[worker] class Worker(
     webUi.stop()
     metricsSystem.stop()
   }
+
+  private def trimFinishedExecutorsIfNecessary(): Unit = {
+    if (finishedExecutors.size > retainedExecutors) {
+      finishedExecutors.take(math.max(finishedExecutors.size / 10, 1)).foreach {
```

Comment:
If this really is safe then we should leave a line comment in the code to record our rationale